collections/src/main/scala/strawman/collection/immutable/ChampHashSet.scala

```diff
@@ -0,0 +1,607 @@
+import Hashing.computeHash
+
+import scala.{Any, AnyRef, Array, Boolean, Double, IndexOutOfBoundsException, Int, NoSuchElementException, SerialVersionUID, Serializable, Unit, `inline`, sys, Some, None}
+import scala.Predef.{assert, require, ???}
+
+import java.lang.Integer.bitCount
+import java.lang.Math.ceil
+import java.lang.System.arraycopy
+
+/** This class implements immutable sets using a Compressed Hash-Array Mapped Prefix-tree.
+  * See paper https://michael.steindorfer.name/publications/oopsla15.pdf for more details.
+  *
+  *  @tparam A      the type of the elements contained in this hash set.
+  *
+  *  @author  Michael J. Steindorfer
+  *  @version 2.13
+  *  @since   2.13
+  *  @define Coll `immutable.ChampHashSet`
+  *  @define coll immutable champ hash set
+  */
+@SerialVersionUID(2L)
+private[immutable] sealed case class ChampHashSet[A](val rootNode: SetNode[A], val cachedJavaHashCode: Int, val cachedSize: Int)
+  extends Set[A]
+    with SetOps[A, ChampHashSet, ChampHashSet[A]]
+    with StrictOptimizedIterableOps[A, ChampHashSet, ChampHashSet[A]]
+    with Serializable {
+
+  def iterableFactory = ChampHashSet
+
+  protected[this] def fromSpecificIterable(coll: collection.Iterable[A]): ChampHashSet[A] = fromIterable(coll)
+
+  protected[this] def newSpecificBuilder(): Builder[A, ChampHashSet[A]] = ChampHashSet.newBuilder()
+
+  override def knownSize: Int = cachedSize
+
+  override def iterator(): Iterator[A] = new SetKeyIterator[A](rootNode)
+
+  def contains(element: A): Boolean = rootNode.contains(element, computeHash(element), 0)
+
+  def incl(element: A): ChampHashSet[A] = {
+    val effect = Effect[A]()
+    val elementHash = computeHash(element)
+    val newRootNode = rootNode.updated(element, elementHash, 0, effect)
+
+    if (effect.isModified)
+      ChampHashSet(newRootNode, cachedJavaHashCode + elementHash, cachedSize + 1)
+    else this
+  }
+
+  def excl(element: A): ChampHashSet[A] = {
+    val effect = Effect[A]()
+    val elementHash = computeHash(element)
+    val newRootNode = rootNode.removed(element, elementHash, 0, effect)
+
+    if (effect.isModified)
+      ChampHashSet(newRootNode, cachedJavaHashCode - elementHash, cachedSize - 1)
+    else this
+  }
+
+  override def empty: ChampHashSet[A] = ChampHashSet.empty
+
+  override def tail: ChampHashSet[A] = this - head
+
+  override def init: ChampHashSet[A] = this - last
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case set: ChampHashSet[A] =>
+        (this eq set) ||
+          (this.cachedSize == set.cachedSize) &&
+            (this.cachedJavaHashCode == set.cachedJavaHashCode) &&
+            (this.rootNode == set.rootNode)
+      case set: Set[A] =>
+        (this eq set) ||
+          (set canEqual this) &&
+            (toIterable.size == set.size) &&
+            (this subsetOf set)
+      case _ => false
+    }
+
+  override def hashCode(): Int = collection.Set.unorderedHash(toIterable, "Set".##)
```

Comment:
This is not necessary.

Reply:
Agree. I'm removing it here, but typically when overriding `equals` I also add an explicit override of `hashCode` to avoid hash-code contract related issues. This is also advised in http://www.artima.com/pins1ed/object-equality.html.

Reply:
Done.